File: src/main/java/net/serlith/purpur/commands/RamCommand.java
package net.serlith.purpur.commands;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import net.serlith.purpur.PurpurBars;
import net.serlith.purpur.configs.RootConfig;
import net.serlith.purpur.tasks.RamBarTask;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.PluginIdentifiableCommand;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class RamCommand extends Command implements PluginIdentifiableCommand {

    private final PurpurBars plugin;
    private final List<String> empty = List.of();

    public RamCommand(PurpurBars plugin) {
        super("ram");
        this.plugin = plugin;

        this.setPermission("purpurbars.check.ram");
        this.setUsage("/ram");
        this.setDescription("Displays server RAM usage in the chat");

        this.permissionMessage(plugin.getPrefix().append(RootConfig.MESSAGES._NO_PERMISSION));
        this.plugin.getServer().getCommandMap().register(this.plugin.getNamespace(), this);
    }

    @Override
    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
        RamBarTask ramTask = RamBarTask.getInstance(this.plugin);
        RootConfig.FORMAT.RAM.OUTPUT.stream().map(i -> MiniMessage.miniMessage().deserialize(i,
                Placeholder.component("allocated", ramTask.format(ramTask.getAllocated())),
                Placeholder.component("used", ramTask.format(ramTask.getUsed())),
                Placeholder.component("xmx", ramTask.format(ramTask.getXmx())),
                Placeholder.component("xms", ramTask.format(ramTask.getXms())),
                Placeholder.component("bar", this.createProgressBar(ramTask.getUsed(), ramTask.getXmx())),
                Placeholder.unparsed("percent", "%d%%".formatted((int) ramTask.getPercent() * 100))
                )).forEach(sender::sendMessage);
        return true;
    }

    private Component createProgressBar(long used, long max) {
        long usedLength = (RootConfig.FORMAT.RAM.USAGE_BAR.LENGTH * used) / max;
        long unusedLength = RootConfig.FORMAT.RAM.USAGE_BAR.LENGTH - usedLength;

        return Component.text(RootConfig.FORMAT.RAM.USAGE_BAR.CHARS.START, RootConfig.FORMAT.RAM.USAGE_BAR.COLOR._BORDER)
                .append(Component.text(RootConfig.FORMAT.RAM.USAGE_BAR.CHARS.BAR.repeat((int) usedLength), RootConfig.FORMAT.RAM.USAGE_BAR.COLOR._USED))
                .append(Component.text(RootConfig.FORMAT.RAM.USAGE_BAR.CHARS.BAR.repeat((int) unusedLength), RootConfig.FORMAT.RAM.USAGE_BAR.COLOR._UNUSED))
                .append(Component.text(RootConfig.FORMAT.RAM.USAGE_BAR.CHARS.END, RootConfig.FORMAT.RAM.USAGE_BAR.COLOR._BORDER));
    }

    @Override
    public @NotNull List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, String[] args) throws IllegalArgumentException {
        return this.empty;
    }

    @Override
    public @NotNull Plugin getPlugin() {
        return this.plugin;
    }
}

File: src/main/kotlin/net/serlith/purpur/commands/RamCommand.kt
package net.serlith.purpur.commands

import net.kyori.adventure.text.ComponentLike
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import net.serlith.purpur.PurpurBars
import net.serlith.purpur.tasks.RamBarTask
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.command.PluginIdentifiableCommand
import org.bukkit.plugin.Plugin

class RamCommand (
    private val plugin: PurpurBars,
) : Command("ram"), PluginIdentifiableCommand {
    private val empty = emptyList<String>()

    init {
        this.permission = "purpurbars.check.ram"
        this.usage = "/ram"
        this.description = "Displays server RAM usage in the chat"

        this.permissionMessage(MiniMessage.miniMessage().deserialize(this.plugin.prefix + this.plugin.mainConfigManager.messages.noPermission))
        this.plugin.server.commandMap.register(this.plugin.namespace, this)
    }

    override fun execute(
        sender: CommandSender,
        commandLabel: String,
        args: Array<out String>?
    ): Boolean {
        val rambar = RamBarTask.instance(this.plugin)
        this.plugin.mainConfigManager.ram.output.map {
            MiniMessage.miniMessage().deserialize(it,
                Placeholder.component("allocated", rambar.format(rambar.allocated)),
                Placeholder.component("used", rambar.format(rambar.used)),
                Placeholder.component("xmx", rambar.format(rambar.xmx)),
                Placeholder.component("xms", rambar.format(rambar.xms)),
                Placeholder.component("bar", this.createProgressBar(rambar.used, rambar.xmx)),
                Placeholder.unparsed("percent", "${(rambar.percent * 100).toInt()}%"),
            )
        }.forEach { sender.sendMessage(it) }
        return true
    }

    private fun createProgressBar(used: Long, max: Long): ComponentLike {
        val usedLength = (this.plugin.mainConfigManager.ram.length * used) / max
        val unusedLength = this.plugin.mainConfigManager.ram.length - usedLength

        val charBar = this.plugin.mainConfigManager.ram.char.bar
        val charStart = this.plugin.mainConfigManager.ram.char.start
        val charEnd = this.plugin.mainConfigManager.ram.char.end

        val usedColor = this.plugin.mainConfigManager.ram.color.used
        val unusedColor = this.plugin.mainConfigManager.ram.color.unused
        val usedBar = "<color:${usedColor}>${charBar.repeat(usedLength.toInt())}</color:${usedColor}>"
        val unusedBar = "<color:${unusedColor}>${charBar.repeat(unusedLength.toInt())}</color:${unusedColor}>"

        val borderColor = this.plugin.mainConfigManager.ram.color.border
        return MiniMessage.miniMessage().deserialize(
            "<color:${borderColor}>$charStart</color:${borderColor}> ${usedBar}${unusedBar} <color:${borderColor}>$charEnd</color:${borderColor}>"
        )
    }

    override fun tabComplete(sender: CommandSender, alias: String, args: Array<out String>?): List<String?> {
        return this.empty
    }

    override fun getPlugin(): Plugin = this.plugin

}